For loops: Allow sharing variables with main program

ajor opened the pull request:
This lets programs such as this run:
```
BEGIN
{
  @map[0] = 1; @map[1] = 2;
  $cnt = 0;
  for ($kv : @map) {
    $cnt++; // Previously would have errored with "Variables defined outside of a for-loop can not be accessed in the loop's scope"
  }
  print($cnt);
}
```

#### Implementation Details
1. Determine which variables need to be shared with the loop callback
2. Pack pointers to them into a context struct
3. Pass pointer to the context struct to the callback function
4. In the callback, override the shared variables so that they read and write through the context pointers instead of directly from their original addresses

Pseudo code for the transformation we apply:

Before:
```
PROBE {
  $str = "hello";
  $not_shared = 2;
  $len = 0;
  @map[11, 12] = "c";
  for ($kv : @map) {
    print($str);
    $len++;
  }
  print($len);
  print($not_shared);
}
```

After:
```
struct ctx_t {
  string *str;
  uint64 *len;
};
PROBE {
  $str = "hello";
  $not_shared = 2;
  $len = 0;
  @map[11, 12] = "c";

  ctx_t ctx { .str = &$str, .len = &$len };
  bpf_for_each_map_elem(@map, &map_for_each_cb, &ctx, 0);

  print($len);
  print($not_shared);
}
long map_for_each_cb(bpf_map *map,
                     const void *key,
                     void *value,
                     void *ctx) {
  $kv = (((uint64, uint64))key, (string)value);
  $str = ((ctx_t*)ctx)->str;
  $len = ((ctx_t*)ctx)->len;

  print($str);
  $len++;
}
```

##### Checklist

N/A: For-loops haven't been released yet and this is basic functionality expected of for-loops so is covered by the existing docs/changelog.
~- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`~
~- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`~
- [x] The new behaviour is covered by tests

jordalgo commented on src/ast/passes/printer.cpp:
```diff
@@ -351,19 +352,27 @@ void Printer::visit(While &while_block)
 }
 
-void Printer::visit(For &for_loop)
+void Printer::visit(For &f)
```
Nit: I know `f` is more terse but I much prefer the variable name `for_loop`.

jordalgo commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2038,6 +2034,62 @@ void SemanticAnalyser::visit(For &f)
+   *     }
+   *
+   * After:
+   *     struct ctx_t {
```
I haven't read the code thoroughly enough yet but is there a possible collision here if there are two (or more) probes that have scratch variables with the same names

jordalgo commented on src/struct.h:
```diff
@@ -176,7 +182,7 @@ class StructManager {
 private:
   std::map<std::string, std::shared_ptr<Struct>> struct_map_;
-  std::unordered_set<std::shared_ptr<Struct>> tuples_;
+  std::unordered_set<std::shared_ptr<Struct>> anonymous_types_;
```
naming nit: maybe make this consistent with the method e.g. `anon_types_` or  `AddAnonymousStruct`

jordalgo commented on tests/semantic_analyser.cpp:
```diff
@@ -197,12 +197,21 @@ void test(std::string_view input, std::string_view expected_ast)
   ast::Printer printer(out);
   printer.print(driver.root.get());
+
+  if (expected_ast[0] == '*' && expected_ast[expected_ast.size() - 1] == '*') {
```
JC, what is this block doing?

jordalgo commented on src/ast/passes/codegen_llvm.h:
```diff
@@ -259,7 +258,11 @@ class CodegenLLVM : public Visitor {
+    llvm::Value *value;
+    llvm::Type *type;
+  };
+  std::map<std::string, VariableLLVM> variables_;
```
Since this is class level is there a case of possible collision where two (or more probes) use the same scratch variable name and reference them in a loop?

ajor replied:
`variables_` is cleared between probes so they won't collide: https://github.com/bpftrace/bpftrace/blob/ca1b216ba92d007c1aa4aa1e46aa6270c8caadaf/src/ast/passes/codegen_llvm.cpp#L2547

jordalgo commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -2471,8 +2471,41 @@ void CodegenLLVM::visit(For &f)
+    ctx_t = b_.GetStructType("ctx_t", ctx_field_types);
+    ctx = b_.CreateAllocaBPF(ctx_t, "ctx");
+
+    for (size_t i = 0; i < ctx_fields.size(); i++) {
```
super duper nit
```suggestion
    for (size_t i = 0; i < ctx_fields.size(); ++i) {
```

ajor replied:
🤮 

sorry, I prefer `i++`

jordalgo commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2038,6 +2034,62 @@ void SemanticAnalyser::visit(For &f)
+   *    write through the context pointers instead of directly from their
+   *    original addresses
+   *
+   * Example transformation with context:
```
I like the example!

jordalgo commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2082,6 +2123,35 @@ void SemanticAnalyser::visit(For &f)
+  for (auto *stmt : *f.stmts) {
+    const auto &live_vars = variable_val_[scope_];
+    vars_referenced.run(*stmt, [&live_vars, &var_set](const auto &var) {
+      if (live_vars.find(var.ident) == live_vars.end())
```
Maybe I'm missing something but why is this check needed?

jordalgo commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2082,6 +2123,35 @@ void SemanticAnalyser::visit(For &f)
+    vars_referenced.run(*stmt, [&live_vars, &var_set](const auto &var) {
+      if (live_vars.find(var.ident) == live_vars.end())
+        return false;
+      if (var_set.find(var.ident) != var_set.end())
```
I think you can safely insert the same string multiple times.

jordalgo commented on src/ast/passes/semantic_analyser.cpp:
```diff
@@ -2082,6 +2123,35 @@ void SemanticAnalyser::visit(For &f)
+  // function as the context parameter.
+  CollectNodes<Variable> vars_referenced;
+  std::unordered_set<std::string> var_set;
+  for (auto *stmt : *f.stmts) {
```
Nit: looks like you could combine these two loops to get referenced and new vars.

ajor replied:
I'd like to delete the "new_vars" loop in hopefully not too long (when #3017 is done), so was keeping them separate to make that easier

jordalgo replied:
Could always add a comment todo with a reference to the issue but that's minor. I won't block.

ajor replied:
The comment on the next for-loop does reference the issue

jordalgo reviewed (commented):
Few small questions and nits but otherwise LGTM.

ajor commented on src/ast/passes/codegen_llvm.h:
```diff
@@ -259,7 +258,11 @@ class CodegenLLVM : public Visitor {
 
-  std::map<std::string, AllocaInst *> variables_;
+  struct VariableLLVM {
+    llvm::Value *value;
```
It's because on this line we we'll have a LoadInst instead of an AllocaInst: https://github.com/bpftrace/bpftrace/pull/3014/files#diff-28926c090177e6efbfb5a7449833e0e863bb226822abb513b7dfc6c388837d87R4257

We're loading a pointer from the ctx tuple and using that to access our variable's memory, instead of going directly through the original AllocaInst.

ajor commented:
Addressed review comments.

Also added a new `AddrSpace::bpf` enum value to represent a pointer to data on the BPF stack (I was previously using `AddrSpace::none`).

jordalgo commented:
Looks like the for-loop semantic analyzer tests are still failing. Fix those and this looks good to go.

ajor commented:
I think I've fixed the test failures now. I was misreading them before - turns out I just needed to add "AS(bpf)" onto the expected results to take into account the new BPF address space I added.

viktormalik reviewed (approved):
LGTM, just one nit.

Also, one test on LLVM 12 is failing (not codegen) and may need some investigation.

viktormalik commented:
We also need to add a CHANGELOG entry since for-loops are now released, right?

ajor commented on src/ast/passes/codegen_llvm.h:
```diff
@@ -273,7 +271,12 @@ class CodegenLLVM : public Visitor {
+    llvm::Value *value;
+    llvm::Type *type;
+  };
+  std::map<std::string, VariableLLVM> variables_;
```
Nope, just historical. I've changed it

ajor commented:
> Also, one test on LLVM 12 is failing (not codegen) and may need some investigation.

This is some craziness! The test shouldn't be any different across LLVM versions, and now I'm also seeing it fail on master on my local machine

ajor commented:
> > Also, one test on LLVM 12 is failing (not codegen) and may need some investigation.
> 
> This is some craziness! The test shouldn't be any different across LLVM versions, and now I'm also seeing it fail on master on my local machine

And the test is passing after a rebase...

viktormalik commented:
> > > Also, one test on LLVM 12 is failing (not codegen) and may need some investigation.
> > 
> > 
> > This is some craziness! The test shouldn't be any different across LLVM versions, and now I'm also seeing it fail on master on my local machine
> 
> And the test is passing after a rebase...

Strange. Could it be caused by a wrong generation of BTF test data? I've a suspicion that something quirky is happening there sometimes.

ajor commented:
> Strange. Could it be caused by a wrong generation of BTF test data? I've a suspicion that something quirky is happening there sometimes.

Yeah, I had a look at the BTF data while the test was failing and it was only showing 3 arguments for the `func_1` function instead of 4. I don't know how that happened though. After a clean build the BTF was showing 4 arguments again.

yump commented:
I am seeing that error message about "Variables defined outside of a for-loop can not be accessed in the loop's scope," that was not supposed to make it into a release, using version 0.21.2 as packaged in Fedora 41, and kernel 6.11.10.

It also happens with the latest v0.21.2 on the releases page:

```
> sudo /tmp/bpftrace -dd -k /tmp/test.bt
[DEBUG] libbpf: loaded kernel BTF from '/sys/kernel/btf/vmlinux'
[DEBUG] libbpf: loaded kernel BTF from '/sys/kernel/btf/vmlinux'

AST after: parser
-------------------
Program
 BEGIN
  =
   map: @map
    int: 0
   int: 1
  =
   map: @map
    int: 1
   int: 2
  =
   variable: $cnt
   int: 0
  for
   decl
    variable: $kv
   expr
    map: @map
   stmts
    variable: $cnt
     ++
  call: print
   variable: $cnt


AST after: ConfigAnalyser
-------------------
Program
 BEGIN
  =
   map: @map
    int: 0 :: [int64]
   int: 1 :: [int64]
  =
   map: @map
    int: 1 :: [int64]
   int: 2 :: [int64]
  =
   variable: $cnt
   int: 0 :: [int64]
  for
   decl
    variable: $kv
   expr
    map: @map
   stmts
    variable: $cnt
     ++
  call: print
   variable: $cnt

/tmp/test.bt:8:5-9: ERROR: Variables defined outside of a for-loop can not be accessed in the loop's scope
    $cnt++; // Previously would have errored with "Variables defined outside of a for-loop can not be accessed in the loop's scope"
    ~~~~
```

ajor commented:
This PR didn't make it into the 0.21 release - it'll be included in the upcoming 0.22.
